doc/README.md

```diff
@@ -0,0 +1,421 @@
+Hence, every repository always uses the same gear table.
+However, to make watermarking attacks harder,
+we use the "app backup gear table key" that gets derived from our main key
+to deterministically compute a gear table using AES CTR to cipher 32 null bytes with a null IV.
+
+TODO: Is it safe to compute the gear table like this? We do derive a dedicated key for this
+      that we don't use for anything else.
+      Still, the IV as well as the encrypted bytes are known to the attacker.
+      Also, it may be be possible to reverse the gear table with chosen plaintext attacks.
+
+## Operations
+
+### Backup
+
+* download all snapshots to get a mapping of all plaintext hashes (chunk IDs) to storage IDs
+* optionally: get a list of all blobs and their storage IDs to confirm they (still) exist
+* if APK backup wasn't disabled, for each app:
+  * get version code and check if we already have an APK with that version code in a snapshot
+  * if APK was backed up already, re-use list of plaintext hashes for current snapshot
+  * if APK was not yet backed up, put it (and its splits) through the chunker
+  * chunks already in the repository are not uploaded again, only their hash recorded
+  * new chunks get compressed, encrypted and hashed to determine their storage ID, then uploaded 
+* for each app (the system allows us to back up):
+  * app data we receive as a tar stream from the system gets chunked
+    * chunks already in the repository are not uploaded again, only their hash recorded
+    * new chunks get compressed, encrypted and hashed to determine their storage ID, then uploaded
+  * remember ordered list of chunk IDs for the app (and its APKs)
+* add newly packed chunks to local index
+* upload consolidated index, remove old index
+* add all apps, their chunk IDs (and related metadata) to new snapshot and upload that
+* at the end prune old snapshots based on retention rules
+
+### Resume interrupted backup
+
+* the mapping from chunk ID to storage ID is only stored in snapshots
+* if backup gets interrupted (e.g. power or network outage), no snapshot gets written
+* due to encryption, storage IDs are not deterministic,  
+  i.e. the same chunk stored two times will have two different storage IDs
+* therefore, we keep a local cache for the mapping of chunk ID to storage ID
+* before starting a backup, we retrieve all uploaded storage IDs 
+  and remove all IDs from our local cache that don't exist
+* when processing new chunks, we check if the chunk ID exists in our local cache 
+  and re-use the existing blob from the mapped storage ID
+* all mappings will be added to snapshot that gets uploaded at the end of backup
+
+### Restore
+
+* download all snapshots and let the user choose one for restore
+* for all apps that were selected for restore:
+  * reinstall app from APKs assembled from fetched chunks (see below for details)
+  * look up chunk IDs for app from snapshot
+  * download blobs for the chunk IDs as specified in snapshot
```

Comment:
[Suggestion/Missing Info] Since the version byte is the only authenticated piece, shouldn't it be verified that the downloaded blobs match their hashes? This was likely already planned but not explicitly mentioned...?

Reply:
> Since the version byte is the only authenticated piece,

Do you know how you got this impression? Maybe we can prevent other from getting that as well.

> shouldn't it be verified that the downloaded blobs match their hashes? This was likely already planned but not explicitly mentioned...?

right, it was left rather implicit. it always needs to happen when reading files, so not sure the high-level procedures is a good place to mention it. I'll find another place.

Reply:
I think I searched the doc for instances of "auth" and the AAD section was the only place it was discussed in detail. As someone not familiar very familiar with the crypto details, to me, that segment sounded like it was justifying why certain things didn't need to be authenticated *at all*.